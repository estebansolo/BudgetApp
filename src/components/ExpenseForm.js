import React, { useState } from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

const ExpenseForm = ({ onSubmit, expense }) => {
	const [description, setDescription] = useState(expense ? expense.description : '');
	const [amount, setAmount] = useState(expense ? (expense.amount / 100).toString() : '');
	const [note, setNote] = useState(expense ? expense.note : '');
	const [createdAt, setCreatedAt] = useState(expense ? moment(expense.createdAt) : moment());
	const [calendarFocused, setCalendarFocused] = useState(false);
	const [error, setError] = useState('');

	return (
		<form
			className="form"
			onSubmit={e => {
				e.preventDefault();
				if (!description || !amount) {
					setError('Please provide description and amount');
				} else {
					setError('');
					onSubmit({
						description,
						amount: parseFloat(amount, 10) * 100,
						createdAt: createdAt.valueOf(),
						note
					});
				}
			}}
		>
			{error && <p className="form__error">{error}</p>}
			<input
				type="text"
				placeholder="Description"
				autoFocus
				value={description}
				className="text-input"
				onChange={e => {
					setDescription(e.target.value);
				}}
			/>
			<input
				type="text"
				placeholder="Amount"
				value={amount}
				className="text-input"
				onChange={e => {
					const amount = e.target.value;
					if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
						setAmount(amount);
					}
				}}
			/>
			<SingleDatePicker
				date={createdAt}
				onDateChange={newCreatedAt => {
					if (newCreatedAt) {
						setCreatedAt(newCreatedAt);
					}
				}}
				focused={calendarFocused}
				numberOfMonths={1}
				onFocusChange={({ focused }) => {
					setCalendarFocused(focused);
				}}
				isOutsideRange={() => false}
			/>
			<textarea
				value={note}
				className="textarea"
				placeholder="Add a note for your expense"
				onChange={e => {
					setNote(e.target.value);
				}}
			/>
			<div>
				<button className="button">Save Expense</button>
			</div>
		</form>
	);
};

export default ExpenseForm;
