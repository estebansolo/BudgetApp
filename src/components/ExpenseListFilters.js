import React, { useState } from 'react';
import { setTextFilter, setEndDate, setStartDate, sortByAmount, sortByDate } from '../actions/filters';
import { connect } from 'react-redux';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';

export const ExpenseListFilters = props => {
	const [calendarFocused, setCalendarFocused] = useState(null);

	const onTextChange = e => {
		props.setTextFilter(e.target.value);
	};

	const onSortChange = e => {
		if (e.target.value === 'date') {
			props.sortByDate();
		} else {
			props.sortByAmount();
		}
	};

	const onDateChanges = ({ startDate, endDate }) => {
		props.setStartDate(startDate);
		props.setEndDate(endDate);
	};

	const onFocusChange = focused => {
		setCalendarFocused(focused);
	};

	return (
		<div className="content-container">
			<div className="input-group">
				<div className="input-group__item">
					<input
						type="text"
						value={props.filters.text}
						onChange={onTextChange}
						className="text-input"
						placeholder="Search Expenses"
					/>
				</div>
				<div className="input-group__item">
					<select onChange={onSortChange} value={props.filters.sortBy} className="select">
						<option value="date">Date</option>
						<option value="amount">Amount</option>
					</select>
				</div>
				<div className="input-group__item">
					<DateRangePicker
						startDate={props.filters.startDate}
						startDateId=""
						endDateId=""
						endDate={props.filters.endDate}
						onDatesChange={onDateChanges}
						focusedInput={calendarFocused}
						onFocusChange={onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
						showClearDates={false}
					/>
				</div>
			</div>
		</div>
	);
};

const mapStateToProps = state => ({
	filters: state.filters
});

const mapDispatchToProps = dispatch => ({
	setTextFilter: text => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: startDate => dispatch(setStartDate(startDate)),
	setEndDate: endDate => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
