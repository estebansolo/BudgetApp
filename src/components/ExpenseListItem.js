import React from 'react';
import { Link } from 'react-router-dom';
import { removeExpense } from '../actions/expenses';
import moment from 'moment';
import { connect } from 'react-redux';
import numeral from 'numeral';

export const ExpenseListItem = ({ description, amount, createdAt, dispatch, id }) => {
	return (
		<div>
			<Link to={`/edit/${id}`}>
				<h3>{description}</h3>
			</Link>
			<p>
				{numeral(amount / 100).format('$0,0.00')}
				-
				{moment(createdAt).format('MMMM Do, YYYY')}
			</p>
			<button
				onClick={() => {
					dispatch(removeExpense({ id }));
				}}
			>
				Remove
			</button>
		</div>
	);
};

export default connect()(ExpenseListItem);
