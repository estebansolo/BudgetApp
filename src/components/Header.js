import React from 'react';
import { connect } from 'react-redux'
import { NavLink } from 'react-router-dom';
import { startLogout } from '../actions/auth' 

export const Header = (props) => (
	<header>
		<h1>Budget App</h1>
		<NavLink exact activeClassName="is-active" to="/dashboard">
			Dashboard
		</NavLink>
		<NavLink activeClassName="is-active" to="/create">
			Create Expense
		</NavLink>
		<NavLink activeClassName="is-active" to="/help">
			Help
		</NavLink>
		<button onClick={props.startLogout}>Logout</button>
	</header>
);

const mapDispatchToProps = (dispatch) => ({
	startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDispatchToProps)(Header)