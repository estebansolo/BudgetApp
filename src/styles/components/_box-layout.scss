.box-layout {
	align-items: center;
	background: $dark-blue;
	display: flex;
	height: 100vh;
	justify-content: center;
	//width: 100vh;
}

.box-layout__boxes {
	display: flex;
	justify-content: space-between;
	margin-top: $l-size;
	> div {
		margin: 0 1rem;
	}

	@media (max-width: $desktop-breakpoint) {
		flex-direction: column;
		align-items: center;
		> div {
			margin: 1rem 0;
		}
	}
}

.box-layout__box {
	background: fade-out($color: #ffffff, $amount: 0.15);
	border-radius: 3px;
	color: $dark-grey;
	padding: $l-size $m-size;
	text-align: center;
	width: 25rem;
	p {
		margin: 0 0 $m-size 0;
	}
}

.box-layout-login {
	text-align: center;
	color: $off-white;
	h1 {
		margin: 0;
	}

	> p {
		margin: 0;
	}
}

.box-layout__title {
	margin: 0 0 $m-size 0;
	line-height: 1;
}
