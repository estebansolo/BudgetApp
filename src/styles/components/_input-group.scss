.input-group {
	display: flex;
	flex-direction: column;
	margin-bottom: $m-size;

	@media (min-width: $desktop-breakpoint) {
		flex-direction: row;
	}
}

.input-group__item {
	margin: 0 $s-size 0 0;

	@media (max-width: $desktop-breakpoint) {
		text-align: center;
		margin-bottom: $s-size;
	}
}
