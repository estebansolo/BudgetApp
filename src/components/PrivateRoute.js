import React from 'react'
import Header from './Header';
import { connect } from 'react-redux'
import { Route, Redirect } from 'react-router-dom'

export const PrivateRoute = ({
    isAuthenticated,
    component: Component,
    ...rest
}) => (
    <Route {...rest} component={(props) => {
        isAuthenticated ? (
            <React.Fragment>
                <Header />
                <Component {...props} />
            </React.Fragment>
        ) : (
            <Redirect to="/" />
        ) 
    }}/>
)

const mapStateToProps = (state) => ({
    isAuthenticated: !!state.auth.uid
})

export default connect(mapStateToProps)(PrivateRoute)