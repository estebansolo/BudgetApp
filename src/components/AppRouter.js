import React from 'react';
import { Router, Route, Switch } from 'react-router-dom';
import { createBrowserHistory } from 'history';

import { ExpenseDashboardPage } from '../pages/ExpenseDashboardPage';
import AddExpensePage from '../pages/AddExpensePage';
import EditExpensePage from '../pages/EditExpensePage';
import LoginPage from '../pages/LoginPage';
import { HelpPage } from '../pages/HelpPage';
import { NotFoundPage } from '../pages/NotFoundPage';
import PrivateRoute from './PrivateRoute'
import PublicRoute from './PublicRoute'

export const history = createBrowserHistory()

const AppRouter = () => (
	<Router history={history}>
		<div>
			<Switch>
				<PublicRoute exact path="/" component={LoginPage}  />
				<PrivateRoute exact path="/dashboard" component={ExpenseDashboardPage} />
				<PrivateRoute exact path="/create" component={AddExpensePage} />
				<PrivateRoute exact path="/edit/:id" component={EditExpensePage} />
				<Route exact path="/help" component={HelpPage} />
				<Route component={NotFoundPage} />
			</Switch>
		</div>
	</Router>
);

export default AppRouter;
