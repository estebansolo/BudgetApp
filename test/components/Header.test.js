// react-test-renderer
import React from 'react';
import { Header } from '../../src/components/Header';
import { shallow } from 'enzyme';
//import ReactShallowRenderer from 'react-test-renderer/shallow';

//test('should render header correctly', () => {
//	const renderer = new ReactShallowRenderer();
//	renderer.render(<Header />);

//	expect(renderer.getRenderOutput()).toMatchSnapshot();
//});

test('should render header correctly', () => {
	const wrapper = shallow(<Header />);
	expect(wrapper).toMatchSnapshot();
	//expect(wrapper.find('h1').text()).toBe('Budget App');
});
