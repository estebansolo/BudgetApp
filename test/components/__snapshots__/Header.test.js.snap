// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header correctly 1`] = `
<header>
  <div
    className="content-container"
  >
    <div
      className="header__content"
    >
      <Link
        className="header__title"
        to="/dashboard"
      >
        <h1>
          Budgets Manager
        </h1>
      </Link>
      <button
        className="button button--link"
        onClick={[Function]}
      >
        Logout
      </button>
    </div>
  </div>
</header>
`;
